Resolve issue with parsing of null bytes (0x00 and 0xFF) before, between, or after a valid tag

bepursuant opened the pull request:
EMV allows for null bytes (0x00 and 0xFF) to exist before, between, and after valid tags - both in constructed tags and outside of them.

This PR adds tests that exhibit an issue with parsing of TLV data that contains padding bytes, and resolves that issue by skipping those 'null' bytes when scanning for a tag. It also adds some comments to clarify the 'magic numbers' used in bitwise operations for parsing multi-byte tags and lengths.

bepursuant committed:
Modified test to add inline and trailing padding (0x00 and 0xFFFF)

bepursuant committed:
When searching for a tag, skip null (0x00 or 0xFF) bytes.
Small changes to clarify var names, and added some comments to clarify parsing procedure.

bepursuant commented:
This update recognizes 0x00 and 0xFF as 'null bytes', which is close to what EMV Book 3 specifies, but not exactly. It states that 0x00 is a 'meaningless byte' that should be skipped when looking for a tag, and also recommends that issuers should not use multi-byte tags starting with 0xFF as some terminals may recognize it as padding.

Should this check follow the actual specification (0x00 only), or should it follow the warning issued in the spec by skipping 0xFF as well?

kspearrin commented on BerTlv/Tlv.cs:
```diff
@@ -104,31 +104,45 @@ private static void ParseTlv(byte[] rawTlv, ICollection<Tlv> result)
                 result.Add(tlv);
 
+                // if this was a constructed tag, parse its value into individual Tlv children as well
                 if(constructedTlv)
```
Please add back the brackets since this is the code style used for the project.

bepursuant replied:
Ha, can do - the empty/bracketless while() above led me to believe it was a looser standard - no worries! Any feedback on whether this should skip both 0x00 and 0xFF, as noted in the PR comment I made? If we can get a decision on the best course there, I can add a new commit.

kspearrin replied:
Honestly, I am not involved in the world of TLV parsing anymore so I cannot speak to what the best course should be. My guess would be to follow the spec.

bepursuant replied:
Added brackets back and changed to skip 0x00 only in newest commit.

bepursuant commented on BerTlv/Tlv.cs:
```diff
@@ -104,31 +104,45 @@ private static void ParseTlv(byte[] rawTlv, ICollection<Tlv> result)
 
+                // bit1 through 5 being 1 indicates that the tag is more than one byte long
+                // recognize this and keep looking until we find the last byte of the tag, indicated by bit8 being 0
+                bool multiByteTag = (rawTlv[i] & 0x1F) == 0x1F;
```
Yes, I was trying to clarify that it's looking at the tag, and follow the lead of the "multiByteLength" var a few lines down. "moreBytes" felt a little ambiguous, so just tried to clean it up a little.

bepursuant replied:
Is this alright or would you prefer it be reverted? I'm going to be implementing some more functionality like recursive tag Find() functionality in the future - given that you're out of the TLV world, would you prefer to maintain ownership and accept PRs, or should I be looking to maintain/release from my own fork going forward? (PS. Thanks for the package!)

bepursuant committed:
Changed null padding to only ignore 0x00 bytes.
Added Parse() helper methods to be more fluent ("Tlv.Parse(string)" vs "Tlv.ParseTlv(string)".
Added test cases for ArgumentExceptions when passed empty string or array.
Renamed a test case for clarity.
